### PR TITLE
feature/Log4jKusto

### PR DESCRIPTION
**Pull request by ag-ramachandran**
- Initial commit
- Sample log4j plugin
- * Add changes and refactor
- * Add E2E tests
- * Add more unit tests * Fix some comments
- * Add license and annotate files with licenses
- * Code coverage and fix code smells
- Create maven.yml
- * Updated dependabot
- * Updated tests
- * Updated workflow file
- * Updated readme
- * Updated package 
- * Add proxy tests
- Rename to azure-kusto-log4j

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+Kusto appender for Log4j.
+
+Motivation and usage.
+----------------------
+Log4j2 is widely used as logging tool. Kusto implementation
+is used in conjunction with RollingFileAppender with KustoStrategy.
+The key reason for using a strategy is to have redundancy in storage
+of logs and re-transmit the log files.
+
+To provide data transmission redundancy , the rolled over log files are
```
Comment (yihezkel) [inline README.md]:
```suggestion
To provide data transmission redundancy, the rolled over log files are
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+transmitted to Kusto. Transmission of the files are attempted 3 times
+with a configured time window
+
+
+Adding appender to log4j.properties
+----------------------
+The key parameters for rolling file are as documented in
+the [Rolling file log4j configuration](https://logging.apache.org/log4j/2.x/manual/appenders.html#RollingFileAppender)
+
+- fileName : The file name where the log files will be written locally. This is a fully qualified path and not a
```
Comment (yihezkel) [inline README.md]:
```suggestion
- fileName: The file name where the log files will be written locally. This is a fully qualified path and not a
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  relative path (e.g C:/logs/logs.log)
+- filePattern : The rolled over file name with pattern. This is a fully qualified path and not a relative path (e.g C:
```
Comment (yihezkel) [inline README.md]:
```suggestion
- filePattern: The rolled over file name with pattern. This is a fully qualified path and not a relative path (e.g C:
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  /logs/logs-%d{yyyy-MM-dd-hh-mm-ss}-%i.log)
+
+Configurations for using the Kusto log4j appender is as follows
+
+- KustoStrategy
+    - **clusterPath**: Ingest URL.Configured using environment variable **LOG4J2_ADX_INGEST_CLUSTER_URL**
```
Comment (yihezkel) [inline README.md]:
```suggestion
    - **clusterPath**: Ingest URL. Configured using environment variable **LOG4J2_ADX_INGEST_CLUSTER_URL**
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    - **appId**: Service principal application id.Configured using environment variable **LOG4J2_ADX_APP_ID**
```
Comment (yihezkel) [inline README.md]:
```suggestion
    - **appId**: Service principal application ID. Configured using environment variable **LOG4J2_ADX_APP_ID**
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    - **appKey**: Service principal application secret.Configured using environment variable **LOG4J2_ADX_APP_KEY**
```
Comment (yihezkel) [inline README.md]:
```suggestion
    - **appKey**: Service principal application secret. Configured using environment variable **LOG4J2_ADX_APP_KEY**
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    - **appTenant**: Tenant for the Service principal.Configured using environment variable **LOG4J2_ADX_TENANT_ID**
```
Comment (yihezkel) [inline README.md]:
```suggestion
    - **appTenant**: Tenant for the service principal. Configured using environment variable **LOG4J2_ADX_TENANT_ID**
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    - **dbName**: Database name.Configured using environment variable **LOG4J2_ADX_DB_NAME**
```
Comment (yihezkel) [inline README.md]:
```suggestion
    - **dbName**: Database name. Configured using environment variable **LOG4J2_ADX_DB_NAME**
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                    flushImmediately=""
+                    proxyUrl=""
+                    backOffMinMinutes=""
+                    backOffMaxMinutes=""
+            />
+            <CsvLogEventLayout delimiter="," quoteMode="ALL"/>
+            <Policies>
+              <!-- Recommended size is 4 MB -->
+                <SizeBasedTriggeringPolicy size="4 MB"/>
+                <TimeBasedTriggeringPolicy interval="5" modulate="true"/>
```
Comment (yihezkel) [inline README.md:77]: Include units

File: README.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+How to build:
+----------------------
+In power-shell the following can be set
+
+```sh
+$env:LOG4J2_ADX_DB_NAME="<db-name>"
+$env:LOG4J2_ADX_ENGINE_URL="https://<cluster>.kusto.windows.net"
+$env:LOG4J2_ADX_TENANT_ID="<tenant-id>"                   
+$env:LOG4J2_ADX_INGEST_CLUSTER_URL="https://ingest-<cluster>.kusto.windows.net"
```
Comment (yihezkel) [inline README.md:96]: I believe there was a recent change that allowed us to deterministically move between the ingestion and query endpoints.
Accordingly, we made this change in Java: https://github.com/Azure/azure-kusto-java/pull/268

File: pom.xml [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    <version>2.18.0</version>
+    <artifactId>azure-kusto-log4j</artifactId>
+    <name>Apache Log4j Kusto</name>
+    <description>Kusto appender for Log4j.</description>
+    <properties>
+        <docLabel>Kusto Documentation</docLabel>
+        <projectDir>/azure-kusto-log4j</projectDir>
+        <module.name>org.apache.logging.log4j.kusto</module.name>
+        <maven.doap.skip>true</maven.doap.skip>
+        <awaitility.version>4.2.0</awaitility.version>
```
Comment (yihezkel) [inline pom.xml:41]: Use empty lines and comments to create distinct blocks. For example, version number properties should be in their own distinct block.

File: src/main/java/org/apache/logging/log4j/ext/kusto/Constants.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+package org.apache.logging.log4j.ext.kusto;
+
+public class Constants {
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/Constants.java]: I don't think we want to have 1 class to define all constants in the app.

Since the first 4 (the ones which begin "LOG4J2_ADX_") are all used only by KustoStrategy.java, I suggest they belong there. And the last is used by KustoClientInstance.java, so I suggest putting it there.

I generally prefer `public`, but just to point out that we can make them `protected` vs `public` depending on whether we only want the relevant tests to use them, or foresee them being used more widely.

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A singleton class that does all the work on the Ingestion of data. Is a singleton instance that instantiates the Ingest client
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
 * A singleton class that does all the work to ingest data. It instantiates the Ingest client
```

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+ * and all operations that are performed from the perspective of ingestion
+ * Uses an external library for performing queued ingestion retries and falls back in case of retries exhausted to write data to a backout
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
 * Uses an external library for performing queued ingestion retries and falls back in case retries are exhausted to write data to a backout
```

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+ * directory where the data can be uploaded in case of transient failures manually.
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
 * directory where the data can be manually uploaded in case of transient failures.
```

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+ * To avoid configuration where too many retries can be configured , the retry attempts have been "hard-limited" (opinionated view) at 3
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
 * To avoid configuration where too many retries can be configured, the retry attempts have been "hard-limited" (opinionated view) at 3
```

Comment (yihezkel) [reply src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: Also, what do you mean by, "opinionated view"?

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    private KustoClientInstance(KustoLog4jConfig kustoLog4jConfig) throws URISyntaxException {
+        // default max attempts is 3 !
+        RetryConfig retryConfig = RetryConfig.custom()
+                .intervalFunction(IntervalFunction.ofExponentialBackoff(kustoLog4jConfig.backOffMinSeconds,
+                        IntervalFunction.DEFAULT_MULTIPLIER, kustoLog4jConfig.backOffMaxSeconds))
+                .retryOnException(this::isTransientException)
+                .failAfterMaxAttempts(false)
+                .build();
+        ingestionRetry = RETRY_REGISTRY.retry(INGESTION_RETRIES, retryConfig);
+        ConnectionStringBuilder csb = "".equals(kustoLog4jConfig.appKey) || "".equals(kustoLog4jConfig.appTenant)
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: Use `StringUtils` methods

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+                ? ConnectionStringBuilder.createWithAadManagedIdentity(kustoLog4jConfig.clusterPath,
+                        kustoLog4jConfig.appId)
+                : ConnectionStringBuilder.createWithAadApplicationCredentials(kustoLog4jConfig.clusterPath,
+                        kustoLog4jConfig.appId,
+                        kustoLog4jConfig.appKey, kustoLog4jConfig.appTenant);
+        csb.setClientVersionForTracing(String.format("Kusto.Log4j.Connector:%s", getPackageVersion()));
+        if (kustoLog4jConfig.proxyUrl != null && !kustoLog4jConfig.proxyUrl.trim().equals("")) {
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: Use `StringUtils` methods

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+            HttpClientProperties proxy = HttpClientProperties.builder().proxy(HttpHost.create(kustoLog4jConfig.proxyUrl)).build();
+            LOGGER.warn("Using proxy : {} ", kustoLog4jConfig.proxyUrl);
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: info level

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+            ingestClient = IngestClientFactory.createClient(csb, proxy);
+        } else {
+            ingestClient = IngestClientFactory.createClient(csb);
+        }
+        LOGGER.info("Using database : {}  & table {} to ingest logs", kustoLog4jConfig.dbName,
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
        LOGGER.info("Ingesting logs to database : {} & table {}", kustoLog4jConfig.dbName,
```

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+                kustoLog4jConfig.tableName);
+        ingestionProperties = new IngestionProperties(kustoLog4jConfig.dbName, kustoLog4jConfig.tableName);
+        ingestionProperties.setFlushImmediately(kustoLog4jConfig.flushImmediately);
+        if (kustoLog4jConfig.logTableMapping != null && kustoLog4jConfig.mappingType != null
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: Use `StringUtils` methods

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+                && !"".equals(kustoLog4jConfig.logTableMapping.trim()) &&
+                !"".equals(kustoLog4jConfig.mappingType.trim())) {
+            LOGGER.error("Using mapping {}  of type  {} ", kustoLog4jConfig.logTableMapping,
+                    kustoLog4jConfig.mappingType);
+            IngestionMapping.IngestionMappingKind mappingType = "json".equalsIgnoreCase(kustoLog4jConfig.mappingType) ? JSON : CSV;
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: Better to enrich the `IngestionMappingKind(String kustoValue)` constructor to accept can insensitive input, and use that.

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+            ingestionProperties.getIngestionMapping()
+                    .setIngestionMappingReference(kustoLog4jConfig.logTableMapping, mappingType);
+        }
+    }
+
+    /**
+     * The instance is created with the config created
```
Comment (yihezkel) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
     * The instance is created based on the KustoLog4jConfig
```

**Review by yihezkel — CHANGES_REQUESTED** [review]
Looks great!

Though I didn't have a chance to go through everything, I have a few suggestions.

File: .github/dependabot.yml [hunk 24]
```diff
@@ -0,0 +1,6 @@
+version: 2
+updates:
+  - package-ecosystem: maven
+    directory: /
+    schedule:
+      interval: weekly
```
Comment (asaharn) [inline .github/dependabot.yml:6]: image
New line is required at the end of file here. It can lead to unexpected behavior when using other tools on the file. 
for ref: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log

File: .travis.yml [hunk 25]
```diff
@@ -0,0 +1,6 @@
+language: java
+jdk:
+  - oraclejdk8
+cache:
+  directories:
+    - $HOME/.m2
```
Comment (asaharn) [inline .travis.yml]: Same for this file. a newline character

File: codecov.yml [hunk 26]
```diff
@@ -0,0 +1,8 @@
+coverage:
+  status:
+    project:
+      default:
+        informational: true
+    patch:
+      default:
+        informational: true
```
Comment (asaharn) [inline codecov.yml]: New line here.

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+    IngestionResult ingestLogs(String filePath) throws IngestionClientException, IngestionServiceException,
+            IOException {
+        try (InputStream inputStream = Files.newInputStream(Paths.get(filePath))) {
+            StreamSourceInfo streamSourceInfo = new StreamSourceInfo(inputStream);
+            return ingestClient.ingestFromStream(streamSourceInfo, ingestionProperties);
+        }
+    }
+
+    /**
+     * In the case the ingestion fails after retries , it is moved to a backout directory that is in the same folder as the path
```
Comment (asaharn) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]:
```suggestion
     * In case, if ingestion fails after retries, it is moved to a backout directory that is in the same folder as the path

```

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoFlushAction.java [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+        this.ingestComplete = false;
+    }
+
+    @Override
+    public boolean execute() throws IOException {
+        boolean execute = delegate.execute();
+        if (execute) {
+            KustoClientInstance.getInstance().ingestRolledFile(fileName);
+        }
+        // reaches here on completion , else IOException gets thrown
```
Comment (asaharn) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoFlushAction.java:29]:
```suggestion

```

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoStrategy.java [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+                config.getStrSubstitutor(),
+                kustoLog4jConfig);
+    }
+
+    private static String getOrEnvVar(String value, String envVarName) {
+        return value != null && value.trim().length() > 0 ? value : System.getenv(envVarName);
+    }
+
+    /**
+     * @param manager The rolling file manager , for example a rolling file which renames files for archival
```
Comment (asaharn) [inline src/main/java/org/apache/logging/log4j/ext/kusto/KustoStrategy.java]:
```suggestion
     * @param manager The rolling file manager, for example a rolling file which renames files for archival

```

File: src/main/resources/app.properties [hunk 30]
```diff
@@ -0,0 +1 @@
+version=${project.version}
```
Comment (asaharn) [inline src/main/resources/app.properties]: Newline required as per github.

File: src/test/resources/delegate.log [hunk 31]
```diff
@@ -0,0 +1 @@
+<Empty>
```
Comment (asaharn) [inline src/test/resources/delegate.log:1]: Is this file needed 🤔

File: src/test/java/org/apache/logging/log4j/ext/kusto/log4j2.component.properties [hunk 32]
```diff
@@ -0,0 +1 @@
+log4j2.configurationFactory=org.apache.logging.log4j.kusto.KustoLog4jTestConfigurationFactory
```
Comment (asaharn) [inline src/test/java/org/apache/logging/log4j/ext/kusto/log4j2.component.properties]:
```suggestion
log4j2.configurationFactory=org.apache.logging.log4j.kusto.KustoLog4jTestConfigurationFactory

```

File: src/test/resources/delegate.log (hunk 31 above)
Comment (asaharn) [inline src/test/resources/delegate.log:1]:
```suggestion
<Empty>

```

File: .github/workflows/maven.yml [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+          distribution: 'temurin'
+          cache: maven
+      - name: Build & with Maven
+        env:
+          LOG4J2_ADX_DB_NAME: ${{ secrets.LOG4J2_ADX_DB_NAME }}
+          LOG4J2_ADX_TENANT_ID: ${{ secrets.LOG4J2_ADX_TENANT_ID }}
+          LOG4J2_ADX_INGEST_CLUSTER_URL: ${{ secrets.LOG4J2_ADX_INGEST_CLUSTER_URL }}
+          LOG4J2_ADX_APP_ID: ${{ secrets.LOG4J2_ADX_APP_ID }}
+          LOG4J2_ADX_APP_KEY: ${{ secrets.LOG4J2_ADX_APP_KEY }}
+        run: mvn -B test --file pom.xml
```
Comment (asaharn) [inline .github/workflows/maven.yml]:
```suggestion
        run: mvn -B test --file pom.xml
        
```

File: src/test/java/org/apache/logging/log4j/ext/kusto/log4j2.component.properties (hunk 32 above)
Comment (ag-ramachandran) [reply src/test/java/org/apache/logging/log4j/ext/kusto/log4j2.component.properties]: @asaharn some of the new line requirements may not be needed as they are used within tests

File: src/test/resources/delegate.log (hunk 31 above)
Comment (ag-ramachandran) [reply src/test/resources/delegate.log:1]: > Looks great!
> 
> Though I didn't have a chance to go through everything, I have a few suggestions.

@yihezkel thanks for the elaborate review. Fixed the review comments and tested. Kindly re-review when you get time. Thanks very much

File: src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java (hunk 27 above)
Comment (ag-ramachandran) [reply src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java]: @asaharn Thanks so much for the review , fixed most of the comments where applicable. Kindly re-review when you get a chance

File: README.md [hunk 34]
```diff
@@ -0,0 +1,125 @@
+Kusto appender for Log4j.
+
+Motivation and usage.
+----------------------
+Log4j2 is widely used as logging tool. Kusto implementation
```
Comment (yihezkel) [inline README.md:5]: is a widely used logging

File: README.md [hunk 35]
```diff
@@ -0,0 +1,125 @@
...
+
+Motivation and usage.
+----------------------
+Log4j2 is widely used as logging tool. Kusto implementation
+is used in conjunction with RollingFileAppender with KustoStrategy.
+The key reason for using a strategy is to have redundancy in storage
+of logs and re-transmit the log files.
+
+To provide data transmission redundancy, the rolled over log files are
+transmitted to Kusto. Transmission of the files are attempted 3 times
```
Comment (yihezkel) [inline README.md:11]: is

File: README.md [hunk 36]
```diff
@@ -0,0 +1,125 @@
...
+with a configured time window
+
+
+Adding appender to log4j.properties
+----------------------
+The key parameters for rolling file are as documented in
```
Comment (yihezkel) [inline README.md:17]: are documented

File: README.md [hunk 37]
```diff
@@ -0,0 +1,125 @@
...
+the [Rolling file log4j configuration](https://logging.apache.org/log4j/2.x/manual/appenders.html#RollingFileAppender)
+
+- fileName: The file name where the log files will be written locally. This is a fully qualified path and not a
```
Comment (yihezkel) [inline README.md:20]: absolute

File: README.md [hunk 38]
```diff
@@ -0,0 +1,125 @@
...
+  relative path (e.g C:/logs/logs.log)
+- filePattern: The rolled over file name with pattern. This is a fully qualified path and not a relative path (e.g C:
```
Comment (yihezkel) [inline README.md:22]: absolute

File: README.md [hunk 39]
```diff
@@ -0,0 +1,125 @@
...
+
+- KustoStrategy
+    - **clusterIngestUrl**: Ingest URL. Configured using environment variable **LOG4J2_ADX_INGEST_CLUSTER_URL**
+    - **appId**: Service principal application id. Configured using environment variable **LOG4J2_ADX_APP_ID**
+    - **appKey**: Service principal application secret. Configured using environment variable **LOG4J2_ADX_APP_KEY**
+    - **appTenant**: Tenant for the Service principal. Configured using environment variable **LOG4J2_ADX_TENANT_ID**
+    - **dbName**: Database name. Configured using environment variable **LOG4J2_ADX_DB_NAME**
+    - **tableName**: Table name for ingesting the logs
+    - **logTableMapping**: Mapping defined in the database to map the log data
+    - **mappingType**: json (or) csv is currently supported. Defaults to **_csv_**
```
Comment (yihezkel) [inline README.md:35]: or

File: README.md [hunk 40]
```diff
@@ -0,0 +1,125 @@
...
+Configurations for using the Kusto log4j appender is as follows
+
+- KustoStrategy
+    - **clusterIngestUrl**: Ingest URL. Configured using environment variable **LOG4J2_ADX_INGEST_CLUSTER_URL**
+    - **appId**: Service principal application id. Configured using environment variable **LOG4J2_ADX_APP_ID**
+    - **appKey**: Service principal application secret. Configured using environment variable **LOG4J2_ADX_APP_KEY**
+    - **appTenant**: Tenant for the Service principal. Configured using environment variable **LOG4J2_ADX_TENANT_ID**
+    - **dbName**: Database name. Configured using environment variable **LOG4J2_ADX_DB_NAME**
+    - **tableName**: Table name for ingesting the logs
+    - **logTableMapping**: Mapping defined in the database to map the log data
```
Comment (yihezkel) [inline README.md:34]: Mapping defined in the database to map the log data's fields to the table's columns.

File: README.md [hunk 41]
```diff
@@ -0,0 +1,125 @@
...
+    - **mappingType**: json (or) csv is currently supported. Defaults to **_csv_**
+    - **flushImmediately**: Boolean indicator to flush the logs immediately. Defaults to **_false_**. Note that making
+      this true may cause additional load on the cluster
+    - **proxyUrl**: Proxy url in case application is hosted behind a proxy
+
+- To attempt retries in case of ingestion failures, retransmission is attempted with the following configuration. 3
```
Comment (yihezkel) [inline README.md:40]: following configuration:

File: README.md [hunk 42]
```diff
@@ -0,0 +1,125 @@
...
+  retries are attempted to ingest the logs. In the event that the file cannot be ingested it gets moved to the backout
```
Comment (yihezkel) [inline README.md:41]: If the

File: README.md [hunk 43]
```diff
@@ -0,0 +1,125 @@
...
+  directory in the same path defined in fileName
+
+    - **backOffMinMinutes**: Min minutes to back off in the event that ingestion fails
```
Comment (yihezkel) [inline README.md:44]: if ingestion

File: README.md [hunk 44]
```diff
@@ -0,0 +1,125 @@
...
+    - **backOffMaxMinutes**: Max minutes to back off in the event that ingestion fails
```
Comment (yihezkel) [inline README.md:45]: if ingestion

File: README.md [hunk 45]
```diff
@@ -0,0 +1,125 @@
...
+$env:LOG4J2_ADX_APP_KEY="<app-key>" 
+```
+followed by 
+
+```mvn clean compiler:compile compiler:testCompile surefire:test```
+
+If you are a maven user, maven dependency plugin can resolve the dependencies. 
+Note : The library uses resilience4j which is brought in automatically by the dependencies to perform retries.
+
+Maven co-ordinates:
```
Comment (yihezkel) [inline README.md:107]: What do you mean by, "co-ordinates"?

File: README.md [hunk 46]
```diff
@@ -0,0 +1,125 @@
...
+To use the library in an application, add the following dependency in maven
+```xml
+<dependency>
+    <groupId>com.microsoft.azure.kusto</groupId>
+    <artifactId>azure-kusto-log4j</artifactId>
+    <version>1.0.0</version>
+</dependency>
+```
+The library expects that `log4j-core` is provided as a dependency in the application. This needs to be included as a 
+dependency, this provides flexibility in using a custom log4j version core library in the application.
```
Comment (yihezkel) [inline README.md:118]: providing flexibility to use

File: pom.xml (hunk 11 above)
Comment (yihezkel) [reply pom.xml:41]: Reminder

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 47]
```diff
@@ -0,0 +1,187 @@
...
+import java.nio.file.Paths;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A singleton class that does all the work to ingest data. It instantiates the Ingest client and all operations that
+ * are performed from the perspective of ingestion. Uses an external library for performing queued ingestion retries
+ * and falls back in case of retries are exhausted to write data to a backout directory where the data can be manually
+ * uploaded in case of transient failures manually. To avoid configuration where too many retries can be configured,
+ * the retry attempts have been "hard-limited" (opinionated view) at 3 by coding it in the application. The time gap
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:45]: What do you mean by "opinionated view"?

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 48]
```diff
@@ -0,0 +1,187 @@
...
+ * between retries are however configurable.
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:46]: is

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 49]
```diff
@@ -0,0 +1,187 @@
...
+        }
+        LOGGER.info("Ingesting logs to database : {} & table {}", kustoLog4jConfig.dbName,
+                kustoLog4jConfig.tableName);
+        ingestionProperties = new IngestionProperties(kustoLog4jConfig.dbName, kustoLog4jConfig.tableName);
+        ingestionProperties.setFlushImmediately(kustoLog4jConfig.flushImmediately);
+        boolean useMapping = StringUtils.isNotEmpty(kustoLog4jConfig.logTableMapping) && StringUtils.isNotEmpty(kustoLog4jConfig.mappingType);
+        if (useMapping) {
+            LOGGER.error("Using mapping {}  of type  {} ", kustoLog4jConfig.logTableMapping,
+                    kustoLog4jConfig.mappingType);
+            IngestionMapping.IngestionMappingKind mappingType = "json".equalsIgnoreCase(kustoLog4jConfig.mappingType) ? JSON : CSV;
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:92]: Compare to Java enum defined in the SDK

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 50]
```diff
@@ -0,0 +1,187 @@
...
+        }
+    }
+
+    /**
+     * The instance that is created with the config is returned for use in parts of the application
+     *
+     * @return The KustoClientInstance that will be operated on in the rest of the application
+     */
+    static KustoClientInstance getInstance() {
+        return ATOMIC_INSTANCE.get();
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:124]: I would expect this to ensure not null, and if it is, populate the singleton.

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 51]
```diff
@@ -0,0 +1,187 @@
...
+
+    /**
+     * In case ingestion fails after retries, the logfile is moved to a backout directory that is in the same folder
+     * as the path where the log file is being processed from (this allows for manual retry)
+     * @param filePath The file that failed processing
+     */
+    void backOutFile(String filePath) {
+        LOGGER.warn("Ingestion failed post retries for file {}. Attempting to move this file to backout", filePath);
+        Path pathOfFile = Paths.get(filePath);
+        String targetDirectory = String.format("%s%sbackout%s", pathOfFile.getParent(), File.separator, File.separator);
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:151]: Consider `Paths.resolve()` to combine dirs

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 52]
```diff
@@ -0,0 +1,187 @@
...
+    private boolean isTransientException(Throwable exception) {
+        Throwable innerException = exception.getCause();
+        return !(innerException instanceof KustoDataExceptionBase &&
+                ((KustoDataExceptionBase) innerException).isPermanent());
+    }
+
+    private static String getPackageVersion() {
+        try {
+            Properties props = new Properties();
+            try (InputStream versionFileStream = KustoClientInstance.class.getResourceAsStream("/app.properties")) {
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:179]: Make sure this works in both Windows and Unix

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoLog4jConfig.java [hunk 53]
```diff
@@ -0,0 +1,41 @@
...
+
+    final String clusterIngestUrl;
+    final String appId;
+    final String appKey;
+    final String appTenant;
+    final String dbName;
+    final String tableName;
+    final String logTableMapping;
+    final String mappingType;
+
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoLog4jConfig.java:15]: nit: Inconsistent newlines

File: src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java [hunk 54]
```diff
@@ -0,0 +1,187 @@
...
+
+/**
+ * A singleton class that does all the work to ingest data. It instantiates the Ingest client and all operations that
+ * are performed from the perspective of ingestion. Uses an external library for performing queued ingestion retries
+ * and falls back in case of retries are exhausted to write data to a backout directory where the data can be manually
+ * uploaded in case of transient failures manually. To avoid configuration where too many retries can be configured,
+ * the retry attempts have been "hard-limited" (opinionated view) at 3 by coding it in the application. The time gap
+ * between retries are however configurable.
+ */
+public final class KustoClientInstance {
```
Comment (yihezkel) [inline src/main/java/com/microsoft/azure/kusto/log4j/KustoClientInstance.java:48]: Best to unit test this class

**Review by yihezkel — APPROVED** [review]
Looking great! Just a few requested changes, and we're good :-)
